File: web/apps/app/lib/core/chat/__tests__/chat-load-state.test.ts
import { describe, expect, it } from 'vitest';
import type { Conversation } from '@oriveo/shared';
import {
  resolveChatLoadState,
  isChatLoadSkeleton,
  isChatComposerBlocked,
  CHAT_LOAD_STALL_TIMEOUT_MS,
  type ChatLoadStateInput,
} from '../chat-load-state';

const CONV_ID = 'E89683FC-24C2-4FEC-9CA6-BBBFFB8F8652';

function makeConversation(overrides: Partial<Conversation> = {}): Conversation {
  return {
    id: CONV_ID,
    title: 'Existing conversation',
    hasCustomTitle: false,
    providerID: '22222222-2222-4222-8222-222222222222',
    providerKind: 'anthropic',
    modelID: 'claude-opus-4-7',
    previewText: '',
    estimatedCost: 0,
    isDraft: false,
    remoteMessageCount: 0,
    messages: [],
    draftText: '',
    updatedAt: '2026-06-04T00:00:00.000Z',
    createdAt: '2026-06-04T00:00:00.000Z',
    ...overrides,
  } as Conversation;
}

function makeInput(overrides: Partial<ChatLoadStateInput> = {}): ChatLoadStateInput {
  return {
    requestedConversationId: CONV_ID,
    conversation: makeConversation(),
    hasLoadedSummary: true,
    localLoadFailed: false,
    backfillPhase: 'idle',
    elapsedSinceEnterMs: 0,
    ...overrides,
  };
}

describe('resolveChatLoadState priority chain', () => {
  it('a local read failure outranks everything else', () => {
    expect(
      resolveChatLoadState(
        makeInput({ localLoadFailed: true, backfillPhase: 'running', elapsedSinceEnterMs: 99_999 }),
      ),
    ).toBe('localFailure');
  });

  it('returns empty when there is no target conversation', () => {
    expect(resolveChatLoadState(makeInput({ requestedConversationId: null }))).toBe('empty');
  });

  it('returns deleted when metadata has loaded and the conversation is gone', () => {
    expect(
      resolveChatLoadState(makeInput({ conversation: undefined, hasLoadedSummary: true })),
    ).toBe('deleted');
  });

  it('does not report deleted before metadata has been read', () => {
    expect(
      resolveChatLoadState(makeInput({ conversation: undefined, hasLoadedSummary: false })),
    ).toBe('bootstrapping');
  });

  //   6 
  it('local content outweighs backfilling, timed out and backfill failed', () => {
    const withMessages = makeConversation({
      messages: [{ id: 'm1', role: 'user', text: 'hi', content: 'hi', state: 'delivered', createdAt: '2026-06-04T00:00:00.000Z' }],
    } as Partial<Conversation>);

    for (const overrides of [
      { backfillPhase: 'running' as const },
      { backfillPhase: 'failed' as const },
      { elapsedSinceEnterMs: CHAT_LOAD_STALL_TIMEOUT_MS },
    ]) {
      expect(resolveChatLoadState(makeInput({ conversation: withMessages, ...overrides }))).toBe(
        'content',
      );
    }
  });

  it('a draft conversation has no remote history to wait for, so it goes to empty', () => {
    expect(
      resolveChatLoadState(makeInput({ conversation: makeConversation({ isDraft: true }) })),
    ).toBe('empty');
    expect(
      resolveChatLoadState(makeInput({ conversation: makeConversation({ draftText: 'half written' }) })),
    ).toBe('empty');
  });

  it('a failed backfill goes to stalled', () => {
    expect(resolveChatLoadState(makeInput({ backfillPhase: 'failed' }))).toBe('stalled');
  });

  it('a backfill still running and not timed out goes to backfilling', () => {
    expect(resolveChatLoadState(makeInput({ backfillPhase: 'running' }))).toBe('backfilling');
  });

  it('a backfill that actually observed 0 remote messages is the only path that lets an existing conversation go to empty', () => {
    expect(resolveChatLoadState(makeInput({ backfillPhase: 'succeededEmpty' }))).toBe('empty');
    // In contrast: with no backfill run yet, "the remote copy is empty too" must not be guessed.
    expect(resolveChatLoadState(makeInput({ backfillPhase: 'idle' }))).toBe('bootstrapping');
  });

  it('an observed 0 that contradicts signs of history in the metadata goes to stalled rather than empty', () => {
    // Zero messages only means "no remote history" when the answer really came from the server.
    // A remote read that tries the backend once and falls back to the local cache on
    // failure, and backfill runs exactly on conversations whose cache is empty, so the fallback
    // snapshot is indistinguishable from a genuine zero. Going to empty would render a
    // conversation with complete remote history as an empty one that is ready to chat in.
    const withCount = makeConversation({ remoteMessageCount: 12 });
    expect(
      resolveChatLoadState(makeInput({ conversation: withCount, backfillPhase: 'succeededEmpty' })),
    ).toBe('stalled');

    const withPreview = makeConversation({ previewText: 'we were halfway through last time' });
    expect(
      resolveChatLoadState(makeInput({ conversation: withPreview, backfillPhase: 'succeededEmpty' })),
    ).toBe('stalled');
  });

  it('previewText and remoteMessageCount take no part in the decision', () => {
    const noisy = makeConversation({ previewText: 'the previous message', remoteMessageCount: 66 });
    expect(resolveChatLoadState(makeInput({ conversation: noisy }))).toBe('bootstrapping');
  });
});

//   8 
describe('12s timeout threshold', () => {
  it('the threshold is exactly 12 seconds', () => {
    expect(CHAT_LOAD_STALL_TIMEOUT_MS).toBe(12_000);
  });

  it('a request that was never sent goes stalled at the threshold, while a running backfill keeps showing', () => {
    expect(
      resolveChatLoadState(
        makeInput({ backfillPhase: 'running', elapsedSinceEnterMs: CHAT_LOAD_STALL_TIMEOUT_MS - 1 }),
      ),
    ).toBe('backfilling');
    expect(
      resolveChatLoadState(
        makeInput({ backfillPhase: 'running', elapsedSinceEnterMs: CHAT_LOAD_STALL_TIMEOUT_MS }),
      ),
    ).toBe('backfilling');
  });

  it('stalled is mutually exclusive with bootstrapping and backfilling', () => {
    const stalled = resolveChatLoadState(
      makeInput({ backfillPhase: 'idle', elapsedSinceEnterMs: CHAT_LOAD_STALL_TIMEOUT_MS }),
    );
    expect(stalled).toBe('stalled');
    expect(isChatLoadSkeleton(stalled)).toBe(false);
  });
});

//   5 stalled  
describe('composer gate', () => {
  it('blocks input for stalled, skeleton and local failure alike', () => {
    expect(isChatComposerBlocked('stalled')).toBe(true);
    expect(isChatComposerBlocked('bootstrapping')).toBe(true);
    expect(isChatComposerBlocked('backfilling')).toBe(true);
    expect(isChatComposerBlocked('localFailure')).toBe(true);
  });

  it('restores sending after a successful backfill', () => {
    expect(isChatComposerBlocked('content')).toBe(false);
    expect(isChatComposerBlocked('empty')).toBe(false);
    expect(isChatComposerBlocked('deleted')).toBe(false);
  });

  it('only backfilling and bootstrapping render as a skeleton', () => {
    expect(isChatLoadSkeleton('backfilling')).toBe(true);
    expect(isChatLoadSkeleton('bootstrapping')).toBe(true);
    expect(isChatLoadSkeleton('content')).toBe(false);
    expect(isChatLoadSkeleton('empty')).toBe(false);
  });
});

/**
 * Regression: a non-draft empty conversation was misjudged as stalled (12s of skeleton, then an
 * error card and a permanently disabled composer that retrying could not clear).
 *
 * The triggers are common: Pro users (whose listener is running) and guests never start a
 * backfill, so backfillPhase stays 'idle' and the chain runs all the way to the timeout. Such
 * conversations really exist: deleting the last message through operations-delete leaves
 * isDraft:false / messages:[] / remoteMessageCount:0, and conversations pulled down from the
 * cloud are all marked isDraft:false. The correct behavior is the writable welcome state after 3s.
 */
describe('empty conversation when backfill is skipped (stalled misjudgement regression)', () => {
  it('skipped with no sign of history goes to empty, not to a skeleton and not to stalled', () => {
    const state = resolveChatLoadState(makeInput({ backfillPhase: 'skipped' }));
    expect(state).toBe('empty');
    expect(isChatLoadSkeleton(state)).toBe(false);
    expect(isChatComposerBlocked(state)).toBe(false);
  });

  it('skipped with a non-empty previewText keeps waiting and never falls to an empty state', () => {
    expect(
      resolveChatLoadState(
        makeInput({
          backfillPhase: 'skipped',
          conversation: makeConversation({ previewText: 'we were halfway through last time' }),
        }),
      ),
    ).toBe('bootstrapping');
  });

  it('skipped with remoteMessageCount > 0 keeps waiting and still goes to stalled on timeout', () => {
    const input = makeInput({
      backfillPhase: 'skipped',
      conversation: makeConversation({ remoteMessageCount: 66 }),
    });
    expect(resolveChatLoadState(input)).toBe('bootstrapping');
    expect(
      resolveChatLoadState({ ...input, elapsedSinceEnterMs: CHAT_LOAD_STALL_TIMEOUT_MS }),
    ).toBe('stalled');
  });

  it("idle ", () => {
    expect(resolveChatLoadState(makeInput({ backfillPhase: 'idle' }))).toBe('bootstrapping');
  });
});

File: android/app/src/test/java/ai/oriveo/community/feature/chat/ChatLoadStateTest.kt
package ai.oriveo.community.feature.chat

import ai.oriveo.community.core.model.ChatMessage
import ai.oriveo.community.core.model.ChatMessageState
import ai.oriveo.community.core.model.ChatRole
import ai.oriveo.community.core.model.Conversation
import ai.oriveo.community.core.model.ProviderKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * State-machine invariants that keep the chat screen from showing a loading skeleton forever.
 *
 * The three that matter most: a message arriving dismisses the skeleton immediately, the skeleton
 * has a hard 12s ceiling, and every waiting state blocks the composer.
 */
class ChatLoadStateTest {

    private val convId = "11111111-1111-1111-1111-111111111111"

    private fun conversation(
        messages: List<ChatMessage> = emptyList(),
        isDraft: Boolean = false,
        draftText: String = "",
        previewText: String = "",
        messageCount: Int = 0,
    ) = Conversation(
        id = convId,
        title = "t",
        providerID = "p",
        providerKind = ProviderKind.OpenAI,
        modelID = "m",
        messages = messages,
        isDraft = isDraft,
        draftText = draftText,
        previewText = previewText,
        messageCount = messageCount,
    )

    private fun message() = ChatMessage(
        id = "msg-1",
        role = ChatRole.User,
        text = "hi",
        providerKind = ProviderKind.OpenAI,
        providerName = "OpenAI",
        modelName = "m",
        state = ChatMessageState.Delivered,
    )

    // -- priority chain --

    @Test
    fun `local failure outranks everything`() {
        assertEquals(
            ChatLoadState.LocalFailure,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messages = listOf(message())),
                hasMissingInitialConversation = true,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS,
                localLoadFailed = true,
            ),
        )
    }

    @Test
    fun `missing conversation outranks content`() {
        assertEquals(
            ChatLoadState.Deleted,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messages = listOf(message())),
                hasMissingInitialConversation = true,
            ),
        )
    }

    /** As soon as a local message shows up, the skeleton is dismissed, however long the wait was. */
    @Test
    fun `content wins over every loading state`() {
        assertEquals(
            ChatLoadState.Content,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messages = listOf(message())),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS * 10,
            ),
        )
    }

    @Test
    fun `brand new chat never enters a loading state`() {
        assertEquals(
            ChatLoadState.Empty,
            resolveChatLoadState(
                requestedConversationId = null,
                conversation = null,
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS * 10,
            ),
        )
    }

    @Test
    fun `draft conversation is empty not stalled`() {
        assertEquals(
            ChatLoadState.Empty,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(isDraft = true),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS * 10,
            ),
        )
    }

    /** The skeleton must have a 12s ceiling; no path may show it forever. */
    @Test
    fun `bootstrapping becomes stalled at the 12s ceiling`() {
        assertEquals(12_000L, CHAT_LOAD_STALLED_TIMEOUT_MS)

        assertEquals(
            ChatLoadState.Bootstrapping,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messageCount = 3),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS - 1,
            ),
        )
        assertEquals(
            ChatLoadState.Stalled,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messageCount = 3),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS,
            ),
        )
    }

    /**
     * An empty local read must not win when it collides with "metadata says there is history".
     *
     * `messageCount` and `previewText` are only ever allowed to veto an empty verdict. Landing on
     * Empty here would render a conversation that does have history as an empty one and let the
     * user start chatting right into it (silently dropping the context, and they would likely
     * delete it out of confusion). Better to keep waiting and then stall.
     */
    @Test
    fun `an empty local read does not win over metadata history evidence`() {
        assertEquals(
            ChatLoadState.Stalled,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messageCount = 12),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS,
            ),
        )
        assertEquals(
            ChatLoadState.Stalled,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(previewText = "left off mid-conversation"),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS,
            ),
        )
    }

    // -- stalled / skeleton states block sending --

    @Test
    fun `stalled and skeleton states block sending`() {
        assertTrue(ChatLoadState.Stalled.blocksSending())
        assertTrue(ChatLoadState.Bootstrapping.blocksSending())
        assertTrue(ChatLoadState.LocalFailure.blocksSending())

        assertFalse(ChatLoadState.Content.blocksSending())
        assertFalse(ChatLoadState.Empty.blocksSending())
        assertFalse(ChatLoadState.Deleted.blocksSending())
    }

    // -- an empty non-draft conversation (a stalled-misjudgment regression) --

    /**
     * Regression: a non-draft empty conversation was being misjudged as stalled (skeleton ->
     * error card + composer permanently disabled, and retrying did not help).
     *
     * This kind of conversation is real: its last message was deleted, and nothing in the stored
     * metadata suggests otherwise. The correct behavior is the welcome page with a usable
     * composer; anything else is a clear regression.
     */
    @Test
    fun `an empty conversation without history evidence shows the welcome state`() {
        val state = resolveChatLoadState(
            requestedConversationId = convId,
            conversation = conversation(),
            hasMissingInitialConversation = false,
        )
        assertEquals(ChatLoadState.Empty, state)
        assertFalse(state.blocksSending())
    }

    /** With evidence of history, an empty read must never land on empty: it waits, then stalls. */
    @Test
    fun `history evidence keeps waiting and then stalls`() {
        assertEquals(
            ChatLoadState.Bootstrapping,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(previewText = "left off mid-conversation"),
                hasMissingInitialConversation = false,
            ),
        )
        assertEquals(
            ChatLoadState.Stalled,
            resolveChatLoadState(
                requestedConversationId = convId,
                conversation = conversation(messageCount = 66),
                hasMissingInitialConversation = false,
                elapsedSinceEnterMs = CHAT_LOAD_STALLED_TIMEOUT_MS,
            ),
        )
    }
}
